### ISSUE TITLE
ONNX 1.8 support timeline

### ISSUE THREAD

**Issue by NathanielMcVicar** [onnx/onnx-mlir#534]
Hi all. We're wondering if there is a timeline or plan for moving to ONNX 1.8.X? We've looked into this locally, and unfortunately, unlike the ONNX 1.7 move, there are some substantial changes required, even without moving to new operator versions. There is a separate issue that the version 13 operator set introduces a number of significant semantic changes, but those can be adopted later or not at all. We're happy to help with the patch, but it touches a lot of core areas. Thanks!

**Comment by NathanielMcVicar** [comment]
@chentong319 We have some of the inference operators at 13 locally if it would be helpful to collaborate on this. If not, I appreciate the schedule info. Thanks!

**Comment by gramalingam** [comment]
A related question: is there any plan for supporting multiple ONNX opsets? Any ONNXMLIR-based tool will need to eventually support ONNX models using earlier opsets, instead of just a single (current) opset. Some of the op-changes in ONNX1.8 break compatibility: e.g., the treatment of "axes" has changed for reduction ops like ReduceSum. If onnxmlir migrates to ONNX1.8, instead of throwing away the lowering-support for the earlier form of ReduceSum, it would be better to just add support for the newer form. But this requires some design decisions on how to support it. One solution would be to generate different dialects for each opset. Another would be to have a single dialect, with different ops (e.g., ReduceSum12 and ReduceSum13).

**Comment by chentong319** [comment]
@NathanielMcVicar I will be very glad to collaborate on this. Which operations did you implemented?

**Comment by chentong319** [comment]
@gramalingam As for the multiple versions support, here are our thoughts:
1. We will focus on support "the latest" version. That means we will keep upgrade onnx-mlir following onnx.
2. As for different dialects for different versions, it is difficult for us to maintain the optimization code. For example, we need write rule for all the dialects. If a new lowering is introduced, all the dialects need to be updated.
3. Our current plan is to use converter to transform lower version to the version onnx-mlir is supporting. The good thing with this approach is that compiler code is clean. The disadvantages include that converter code may not be available and it is always guaranteed  (I think in theory) we can write converter for all operations.
4. I think v11 ReduceSum can be converted to v13 ReduceSum by convert the attribute to constant and use it an input. Is that correct?
5. I noticed some code in Build that handles the optional input with default value need to be changed when we move to new version. I might keep those code in certain way so that we can import different version operation if we define operation for them.
@gramalingam @NathanielMcVicar You already have some experience for versioning.  We can discuss and collaborate more on this topic.

**Comment by AlexandreEichenberger** [comment]
Agreeing with @chentong319 opinion.

On supporting multiple ONNX opsets, my hope is that we can handle most of the ops by up converting older ops to a newer version, so that we may have as few operations as possible. This is borne by the experience that the more ways that exist to do something similar, the more likely something is not going to be fully tested when some new optimization is brought online. 

When an operation is too different from prior one, meaning we cannot convert an old one in a newer one, then we will have to maintain two operations, with two distinct names to maintain the past behavior and the newer behavior. 

The idea of two dialects seems to introduce more redundancy to me, maybe there is a way to have it so that most ops are reused, an idea to be maybe pursued a bit further. It might be a solution if we can introduce a version specific dialect, and then use MLIR rules to automatically convert them in the latest version. But having two or more ONNX dialects that needs to be handled separately all the way to their lowering to KRNL and eventually affine/vector/standard, that does not seem attractive at all to me.

**Comment by NathanielMcVicar** [comment]
It looks like the ONNX 1.8 patch is incoming #575. It looks like we have the Expand op, and maybe some others (I need to count) ready to go. Would it make sense for us to start submitting PRs for individual ops once 575 lands? Thanks!

**Comment by caoimhinuibrian** [comment]
@NathanielMcVicar Yes, I think if you have patches ready to go it would be great to submit PRs for them.

**Comment by NathanielMcVicar** [comment]
I'm going to close this, since 1.8 specifically is no longer relevant, and future discussion of supporting multiple opsets can continue in #708.
